oci/layout/oci_dest.go

```diff
@@ -184,23 +193,20 @@ func (d *ociImageDestination) PutManifest(m []byte) error {
 		Architecture: runtime.GOARCH,
 		OS:           runtime.GOOS,
 	}
-	d.index.Manifests = append(d.index.Manifests, desc)
+	d.addManifest(&desc)
 
 	return nil
 }
 
-func ensureDirectoryExists(path string) error {
-	if _, err := os.Stat(path); err != nil && os.IsNotExist(err) {
-		if err := os.MkdirAll(path, 0755); err != nil {
-			return err
+func (d *ociImageDestination) addManifest(desc *imgspecv1.Descriptor){
+	for i, manifest := range d.index.Manifests {
+		if manifest.Digest.String() == desc.Digest.String() {
```

Comment:
Should this really be overwriting based on the manifest digest? If the manifest digest is the same, then the image by definition hasn’t changed (well, apart from the external metadata).

My initial impression is that this should replace images with the same `org.opencontainers.image.ref.name` annotation, if any; does that make sense?

Reply:
> BTW, I guess this is what I am after - https://github.com/containers/image/pull/318/files

Yes, this PR and #318 complement each other; #318 handles the permitted names and the read side, this one handles the write side.

Reply:
@cyphar 

> The image name should not be part of the reference. The reference is the "tag". It doesn't make sense to include the path to the image, because you can move the image to a different path and because there's no cryptographic verification of image references.

Maybe we are just using different terminology? In the context of containers/image, a “reference” is most likely a `types.ImageReference`, a (local-process-only, not necessarily portable to any other environment) pointer to any kind of _single_ image; in this particular case, `ociReference` is a (filesystem directory, `org.opencontainers.image.ref.name` annotation value) pair referring to an image within an index in an on-filesystem OCI-formatted directory. The annotation value is usually called a “tag” or an “image” in the current PRs. The filesystem directory path is an obviously necessary part of `types.ImageReference`, the callers must point to a directory _somehow_, but it is in no way embedded into the contents of the OCI repo.

>
> Please, just use the current scheme. We had so many issues with references inside the OCI spec. I appreciate that the current spec doesn't forbid you from making a decision like that, I seriously would not recommend it.

What do you mean by “the current scheme” / “a decision like that”?

>I think #318 is not doing the right thing.

In what way? AFAICS all it does WRT naming is that it expands the grammar of allowed …`.name` annotation names from a tag-like regexp to the regexp in the OCI standard, without any conversions or normalizations. It is not opinionated in any way, it just allows callers to use the OCI standard as it is written.

(and, sure, with the OCI standard saying nothing about the semantics of the image names, maybe there could be interoperability problems / mismatches in expectation; But that seems to be _also_ to be an argument for containers/image just exposing the raw data and letting the caller deal with whatever the environment and semantics is for that particular image.)

Reply:
@hferentschik 
> > I'd recommend taking a look at https://github.com/openSUSE/umoci's interface to see what I mean (disclosure, I wrote it).
>
> The link does not seem to work. Can you point to the actual interface?

Try https://github.com/openSUSE/umoci .

Reply:
@hferentschik It looks like GitHub screwed up the link. The exact link is https://github.com/openSUSE/umoci/blob/master/oci/casext/refname.go -- which contains all of the umoci's reference handling code (including resolution and replacement of references).

Reply:
> It looks like GitHub screwed up the link. The exact link is https://github.com/openSUSE/umoci/blob/master/oci/casext/refname.go -- which contains all of the umoci's reference handling code (including resolution and replacement of references).

Ok. Looking just at this file it is hard to tell how it is used though. Can you post the top level API call a user of this library would make to put say alpine:3.1 into the index?

Reply:
[See also #323 for discussion on the `oci:`(path, annotation) syntax.)

Reply:
@cyphar 

> Lastly there is a question of spec interpretation. I would expect that if I want to push alpine:3.1 into an OCI index, the top level entry in the index should use 'org.opencontainers.image.ref.name' as name for the image, aka 'alpine' and 'org.opencontainers.image.ref.version' as version, aka 3.1.

This is actually a reasonable point, and AFAICT the OCI spec is silent on this. Is it reasonable / interoperable with umoci to expect users to set …`.ref.name` to unique values (i.e. embed Docker-like version tags in there), or should every consumer be matching on a pair of `.ref.name`+`.ref.version`, or perhaps on even more annotations, because single annotations or subsets of that are not expected to be unique?

Reply:
@hferentschik What do you mean by "put alpine:3.1" in the index? Do you mean a refname of `"alpine:3.1"`? If so, you'd do something like `(casext.Engine).AddReferences(context.Background(), "alpine:3.1", <descriptor that points to the manifest>)`. But I'm not fully sure what you're asking to be honest.

Reply:
@mtrmac 

> > Lastly there is a question of spec interpretation. I would expect that if I want to push alpine:3.1 into an OCI index, the top level entry in the index should use 'org.opencontainers.image.ref.name' as name for the image, aka 'alpine' and 'org.opencontainers.image.ref.version' as version, aka 3.1.

> This is actually a reasonable point, and AFAICT the OCI spec is silent on this. Is it reasonable / interoperable with umoci to expect users to set ….ref.name to unique values (i.e. embed Docker-like version tags in there), or should every consumer be matching on a pair of .ref.name+.ref.version, or perhaps on even more annotations, because single annotations or subsets of that are not expected to be unique?

To be interoperable with `umoci`, `umoci` only looks at `org.opencontainers.image.ref.name`. It internally does support handling of multiple entries with the same name, but the UX doesn't support it (and so it does some not-very-nice things). As for the general concept, I'm not really sure which way is better. While I understand why you'd want to use different `org.opencontainers.image.ref.*` things, I would be careful about adding crucial information to annotations that other implementations might not know about. Do I think this is something that will need to be refined in future OCI updates? Most definitely, the arguments we had about issues like this before 1.0 are pages long.

Also, if you want to inter-operate with Docker (you'd have to ask @runcom to explain how the current implementation is going) then it's likely that you'll have to use `ref.name` only.

Reply:
Great, `.ref.name`-only it is, then.

I was worried about having to invent some SQL-like query language, that would be a pretty bad UX (and a very horrible UX if every OCI consumer defined its own).